### PR TITLE
rgw: sync modules, metadata search

### PR DESCRIPTION
**Pull request by yehudasa**
Make data sync more modular, so that we could add sync modules. An example for such sync module is the "log" module that logs every object that needs to be synced. Other examples (yet to be implemented) are meta indexing module, or backup (to external storage) module.

### CODE REVIEW

File: src/rgw/rgw_sync_module.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  virtual bool supports_data_export() = 0;
+  virtual int create_instance(CephContext *cct, map<string, string>& config, RGWSyncModuleInstanceRef *instance) = 0;
+};
+
+typedef std::shared_ptr<RGWSyncModule> RGWSyncModuleRef;
+
+
+class RGWSyncModulesManager {
+  Mutex lock;
```
Comment (cbodley) [inline src/rgw/rgw_sync_module.h]: unused

**Comment by yehudasa** [comment]
@cbodley addressed you comments, repushed

**Comment by cbodley** [comment]
looks good. is it passing test_multi.py?

**Comment by yehudasa** [comment]
@cbodley yes

**Pull request merged**